hibernate-core/src/main/java/org/hibernate/query/spi/StreamDecorator.java

```diff
@@ -0,0 +1,308 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.query.spi;
+
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.Optional;
+import java.util.Spliterator;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.BinaryOperator;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.IntFunction;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+import java.util.function.ToDoubleFunction;
+import java.util.function.ToIntFunction;
+import java.util.function.ToLongFunction;
+import java.util.stream.Collector;
+import java.util.stream.DoubleStream;
+import java.util.stream.IntStream;
+import java.util.stream.LongStream;
+import java.util.stream.Stream;
+
+import org.hibernate.Incubating;
+
+/**
+ * The {@link StreamDecorator} wraps a Java {@link Stream} and registers a {@code closeHandler}
+ * which is passed further to any resulting {@link Stream}.
+ *
+ * The goal of the {@link StreamDecorator} is to close the underlying {@link Stream} upon
+ * calling a terminal operation.
+ *
+ * @author Vlad Mihalcea
+ * @since 5.4
+ */
+public class StreamDecorator<R> implements Stream<R> {
```

Comment:
Seems another exciting PR from Vlad. I just wanna raise some awareness of the maintenance cost of the current implementation. So suppose next JVM adds some new method to Stream, do we need to modify the class accordingly to sync up? Seems not a concern for JVM maintains backward compatibility by interface default method?

Other than delegator, Spring's AOP might be a good implementation alternative provided JDK exposes some mean to know whether the current method is 'terminal' one or not.

Reply:
@sebersole @NathanQingyangXu I thought of that too, but then I run the tests with Java 13 and there was no issue. I've been using this approach in {FlexyPool](https://github.com/vladmihalcea/flexy-pool) as an alternative to dynamic proxies for the JDBC `Connection`. If there's a method added in a new JDK version, and we need to compile with JDK 8, we can just add the new methods without `@Override` and use Reflection to invoke the method on the delegate.

Another approach is to use Byte Buddy, but there will be many if/else checks to determine which methods we need to call `close` and which require creating a new `StreamDecorator`.

Reply:
@sebersole @NathanQingyangXu Cool. I found why it works on Java 13. It looks like the JDK folks decided to mark the new `Stream` methods using `default`, so that's why the test run. However, we still need to add those so that we can either decorate the `Stream` or call `close` if it's a terminator operation. I'll issue a new commit based on these comments.